File: src/test/java/com/testainers/RedirectResourceTest.java
package com.testainers;

import io.quarkus.test.junit.QuarkusTest;
import io.restassured.RestAssured;
import io.restassured.config.RestAssuredConfig;
import io.restassured.http.ContentType;
import io.restassured.http.Method;
import jakarta.ws.rs.core.HttpHeaders;
import org.junit.jupiter.api.Test;

import java.util.List;

import static io.restassured.RestAssured.given;
import static io.restassured.config.RedirectConfig.redirectConfig;
import static org.hamcrest.Matchers.is;

/**
 * @author Eduardo Folly
 */
@QuarkusTest
public class RedirectResourceTest {

    static final String LOCATION = "https://testainers.com";

    static final String INVALID_SCHEME = "testainers.com";

    static final List<Method> METHODS =
            List.of(Method.GET, Method.POST, Method.PUT,
                    Method.PATCH, Method.DELETE);

    static final RestAssuredConfig CONFIG = RestAssured
            .config()
            .redirect(redirectConfig().followRedirects(false));

    @Test
    public void redirect302() {
        for (Method method : METHODS) {
            given().config(CONFIG)
                   .when()
                   .queryParam("url", LOCATION)
                   .request(method, "/redirect")
                   .then()
                   .statusCode(302)
                   .header(HttpHeaders.LOCATION, LOCATION)
                   .header(HttpHeaders.CONTENT_LENGTH, "0");
        }
    }

    @Test
    public void redirect303() {
        for (Method method : METHODS) {
            given().config(CONFIG)
                   .when()
                   .queryParam("url", LOCATION)
                   .queryParam("code", 303)
                   .request(method, "/redirect")
                   .then()
                   .statusCode(303)
                   .header(HttpHeaders.LOCATION, LOCATION)
                   .header(HttpHeaders.CONTENT_LENGTH, "0");
        }
    }

    @Test
    public void redirect299() {
        for (Method method : METHODS) {
            given().config(CONFIG)
                   .when()
                   .queryParam("url", LOCATION)
                   .queryParam("code", 299)
                   .request(method, "/redirect")
                   .then()
                   .statusCode(500)
                   .contentType(ContentType.TEXT)
                   .body(is("Invalid status code: 299"));
        }
    }

    @Test
    public void redirect499() {
        for (Method method : METHODS) {
            given().config(CONFIG)
                   .when()
                   .queryParam("url", LOCATION)
                   .queryParam("code", 499)
                   .request(method, "/redirect")
                   .then()
                   .statusCode(500)
                   .contentType(ContentType.TEXT)
                   .body(is("Invalid status code: 499"));
        }
    }

    @Test
    public void redirectNoUrl() {
        for (Method method : METHODS) {
            given().config(CONFIG)
                   .when()
                   .request(method, "/redirect")
                   .then()
                   .statusCode(500)
                   .contentType(ContentType.TEXT)
                   .body(is("Invalid URL: null"));
        }
    }

    @Test
    public void redirectEmptyUrl() {
        for (Method method : METHODS) {
            given().config(CONFIG)
                   .when()
                   .queryParam("url", "")
                   .request(method, "/redirect")
                   .then()
                   .statusCode(500)
                   .contentType(ContentType.TEXT)
                   .body(is("Invalid URL Scheme: "));
        }
    }

    @Test
    public void redirectInvalidScheme() {
        for (Method method : METHODS) {
            given().config(CONFIG)
                   .when()
                   .queryParam("url", INVALID_SCHEME)
                   .request(method, "/redirect")
                   .then()
                   .statusCode(500)
                   .contentType(ContentType.TEXT)
                   .body(is("Invalid URL Scheme: " + INVALID_SCHEME));
        }
    }

}

File: src/test/kotlin/com/testainers/RedirectResourceTest.kt
package com.testainers

import io.quarkus.test.junit.QuarkusTest
import io.restassured.http.ContentType
import io.restassured.http.Method
import io.restassured.module.kotlin.extensions.*
import jakarta.ws.rs.core.HttpHeaders
import org.hamcrest.Matchers.*
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

/**
 * @author Eduardo Folly
 */
@QuarkusTest
class RedirectResourceTest : BaseResourceTest() {
    companion object : BaseResourceTest() {
        const val LOCATION = "https://testainers.com"

        @JvmStatic
        fun invalidRedirect(): List<Arguments> =
            argumentGenerator(listOf(299, 400))

        @JvmStatic
        fun requiredUrl(): List<Arguments> =
            argumentGenerator(listOf(null, "", " "))

        @JvmStatic
        fun invalidUrl(): List<Arguments> =
            argumentGenerator(listOf("a{b}", "a:|b"))

        @JvmStatic
        fun invalidScheme(): List<Arguments> =
            argumentGenerator(
                listOf(
                    "testainers.com",
                    "1.8",
                    "a@b",
                    "ftp://teste.com",
                ),
            )

        @JvmStatic
        fun validRedirect(): List<Arguments> =
            argumentGenerator(listOf(null, "", "302", "303"))
    }

    @ParameterizedTest
    @MethodSource("invalidRedirect")
    fun invalid(
        method: Method,
        code: Int,
    ) {
        Given {
            config(config)
        } When {
            queryParam("url", LOCATION)
            queryParam("code", code)
            request(method, "/redirect")
        } Then {
            statusCode(500)
            if (method != Method.HEAD) {
                contentType(ContentType.TEXT)
                body(equalTo("Invalid status code: $code"))
            }
        }
    }

    @ParameterizedTest
    @MethodSource("validRedirect")
    fun valid(
        method: Method,
        code: String?,
    ) {
        Given {
            config(config)
        } When {
            queryParam("url", LOCATION)
            if (!code.isNullOrBlank()) {
                queryParam("code", code)
            }
            request(method, "/redirect")
        } Then {
            statusCode(code?.toIntOrNull() ?: 302)
            header(HttpHeaders.LOCATION, LOCATION)
            if (method != Method.HEAD) {
                header(HttpHeaders.CONTENT_LENGTH, "0")
            }
        }
    }

    @ParameterizedTest
    @MethodSource("requiredUrl")
    fun required(
        method: Method,
        url: String?,
    ) {
        Given {
            config(config)
        } When {
            if (url != null) {
                queryParam("url", url)
            }
            request(method, "/redirect")
        } Then {
            statusCode(500)
            if (method != Method.HEAD) {
                contentType(ContentType.TEXT)
                body(equalTo("URL is required"))
            }
        }
    }

    @ParameterizedTest
    @MethodSource("invalidUrl")
    fun invalid(
        method: Method,
        url: String?,
    ) {
        Given {
            config(config)
        } When {
            if (url != null) {
                queryParam("url", url)
            }
            request(method, "/redirect")
        } Then {
            statusCode(500)
            if (method != Method.HEAD) {
                contentType(ContentType.TEXT)
                body(equalTo("Invalid URL: $url"))
            }
        }
    }

    @ParameterizedTest
    @MethodSource("invalidScheme")
    fun scheme(
        method: Method,
        url: String?,
    ) {
        Given {
            config(config)
        } When {
            if (url != null) {
                queryParam("url", url)
            }
            request(method, "/redirect")
        } Then {
            statusCode(500)
            if (method != Method.HEAD) {
                contentType(ContentType.TEXT)
                body(equalTo("Invalid URL Scheme: $url"))
            }
        }
    }
}
